packages/storage/src/namespace_helpers.rs

```diff
@@ -134,4 +199,109 @@ mod test {
+        assert_eq!(first, (expected_key, b"none".to_vec()));
+    }
+
+    #[test]
+    #[cfg(feature = "iterator")]
+    fn test_range_with_prefix_wrapover() {
+        let mut storage = MockStorage::new();
+        // if we don't properly wrap over there will be issues here (note 255+1 is used to calculate end)
+        let prefix = key_prefix(b"f\xff\xff");
+        let other_prefix = key_prefix(b"f\xff\x44");
+
+        // set some values in this range
+        set_with_prefix(&mut storage, &prefix, b"bar", b"none");
+        set_with_prefix(&mut storage, &prefix, b"snowy", b"day");
+
+        // set some values outside this range
+        set_with_prefix(&mut storage, &other_prefix, b"moon", b"buggy");
+
+        // ensure we get proper result from prefixed_range iterator
+        let mut iter = range_with_prefix(&storage, &prefix, None, None, Order::Descending);
+        let first = iter.next().unwrap();
+        assert_eq!(first, (b"snowy".to_vec(), b"day".to_vec()));
+        let second = iter.next().unwrap();
+        assert_eq!(second, (b"bar".to_vec(), b"none".to_vec()));
+        assert_eq!(iter.next(), None);
+    }
+
+    #[test]
+    #[cfg(feature = "iterator")]
+    fn test_range_with_start_end_set() {
+        let mut storage = MockStorage::new();
+        // if we don't properly wrap over there will be issues here (note 255+1 is used to calculate end)
+        let prefix = key_prefix(b"f\xff\xff");
+        let other_prefix = key_prefix(b"f\xff\x44");
+
+        // set some values in this range
+        set_with_prefix(&mut storage, &prefix, b"bar", b"none");
+        set_with_prefix(&mut storage, &prefix, b"snowy", b"day");
+
+        // set some values outside this range
+        set_with_prefix(&mut storage, &other_prefix, b"moon", b"buggy");
+
+        // make sure start and end are applied properly
+        let res: Vec<KV> =
+            range_with_prefix(&storage, &prefix, Some(b"b"), Some(b"c"), Order::Ascending)
+                .collect();
+        assert_eq!(res.len(), 1);
+        assert_eq!(res[0], (b"bar".to_vec(), b"none".to_vec()));
+
+        // make sure start and end are applied properly
+        let res: Vec<KV> = range_with_prefix(
+            &storage,
+            &prefix,
+            Some(b"bas"),
+            Some(b"sno"),
+            Order::Ascending,
+        )
+        .collect();
+        assert_eq!(res.len(), 0);
+
+        let res: Vec<KV> =
+            range_with_prefix(&storage, &prefix, Some(b"ant"), None, Order::Ascending).collect();
+        assert_eq!(res.len(), 2);
+        assert_eq!(res[0], (b"bar".to_vec(), b"none".to_vec()));
+        assert_eq!(res[1], (b"snowy".to_vec(), b"day".to_vec()));
+    }
+
+    #[test]
+    #[cfg(feature = "iterator")]
+    fn increment_bytes() {
+        assert_eq!(increment_last_byte(b"bob"), b"boc".to_vec());
+        assert_eq!(increment_last_byte(b"fo\xfe"), b"fo\xff".to_vec());
+        assert_eq!(increment_last_byte(b"fo\xff"), b"fp\x00".to_vec());
```

Comment:
Can you also test those?

```rust
assert_eq!(increment_last_byte(b""), b"\x01".to_vec());
assert_eq!(increment_last_byte(b"\xff"), b"\xff\x01".to_vec());
assert_eq!(increment_last_byte(b"\xff\xff\xff"), b"\xff\xff\xff\x01".to_vec());
```

My feeling says they'll fail

Reply:
Ah, and in general we need to increment into the the other direction. Between `fo\xff` and `fp\x00` there is a lot more which we don't want to include:

```
fo\xff
fo\xff\x01
fo\xff\x02
fo\xff\x03
fo\xff\x04
fo\xff\x04abc
...
fo\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff....
fp\x00
```

Reply:
Now I'm unsure what we really want here 🤯

Rename to `fn next_namespace(namespace: &[u8]) -> Vec<u8>` and keep the implementation as is?

Reply:
Yeah, we know that prefixes will never be empty, and the only way to have them pure 0xff is to have one of length 65535 of pure 0xff. I guess we don't handle that case correctly. However, the actual proper return value in any of these cases is to use `None` as an upper bound. I can update it for that. (Option<Vec<u8>>)?

Reply:
> we know that prefixes will never be empty

Right, because even an empty user input is length-prefixed to `0`.

> and the only way to have them pure 0xff is to have one of length 65535 of pure 0xff. I guess we don't handle that case correctly.

If the implementation is solely used for length-prefixed keys, a namespace that only contains 0xff cannot exist. With a little comment like this, we can avoid the `Opltion`ality.

I think the most important thing here is the renaming, to make clear why we can do a few assumptions.

Reply:
I fixed this in 64b21a1

Happy to keep this correct but complex approach, or revert it as you see fit.